aiida/tools/importexport/dbimport/readers.py

```diff
@@ -0,0 +1,328 @@
+
+    @abstractmethod
+    def node_repository(self, uuid: str) -> Folder:
+        """Return a temporary folder containing the contents of the repository for a single node.
+
+        :raises `~aiida.tools.importexport.common.exceptions.CorruptArchive`: If the repository does not exist.
+        """
+
+
+class ReaderJsonZip(ArchiveReaderAbstract):
+    """A reader for a Json zip compressed format."""
+
+    def __init__(self, filename: str, sandbox_in_repo: bool = True, **kwargs: Any):
+        """A writer for zipped archives.
+
+        :param filename: the filename (possibly including the absolute path)
+            of the file on which to export.
+
+        """
+        super().__init__(filename, **kwargs)
+        self._metadata = None
+        self._data = None
+        # a temporary folder used to extract the file tree
+        self._sandbox: Optional[SandboxFolder] = None
+        self._sandbox_in_repo = sandbox_in_repo
+        self._nodes_export_subfolder = NODES_EXPORT_SUBFOLDER
+        self._extracted = False
+
+    @property
+    def file_format_verbose(self) -> str:
+        return 'Zip'
+
+    @property
+    def compatible_export_version(self) -> str:
+        """Return the export version that this reader is compatible with."""
+        return EXPORT_VERSION
+
+    @ensure_within_context
+    def _get_metadata(self):
+        """Retrieve the metadata JSON."""
+        if self._metadata is None:
+            try:
+                self._metadata = json.loads(
+                    zipfile.ZipFile(self.filename, 'r', allowZip64=True).read('metadata.json').decode('utf8')
+                )
+            except zipfile.BadZipfile:
+                raise ValueError('The input file format is not valid (not a zip file)')
+            except KeyError:
+                raise CorruptArchive('required file `metadata.json` is not included')
+        return self._metadata
+
+    @ensure_within_context
+    def _get_data(self):
+        """Retrieve the data JSON."""
+        if self._data is None:
+            try:
+                self._data = json.loads(
+                    zipfile.ZipFile(self.filename, 'r', allowZip64=True).read('data.json').decode('utf8')
+                )
+            except zipfile.BadZipfile:
+                raise ValueError('The input file format is not valid (not a zip file)')
+            except KeyError:
+                raise CorruptArchive('required file `data.json` is not included')
+        return self._data
+
+    @ensure_within_context
+    def _extract(self, path_prefix: str = ''):
+        """Extract repository data to a temporary folder.
+
+        :param path_prefix: Only extract paths starting with this prefix.
+
+        :raises TypeError: if parameter types are not respected
+        """
+        try:
+            with zipfile.ZipFile(self.filename, 'r', allowZip64=True) as handle:
+                for membername in handle.namelist():
```

Comment:
@ltalirz note this, together with the `node_repository` method, is a bit of a performance regression for the current archive format, because you have to iterate through all of `handle.namelist()` each time you want to extract/return a node repository, e.g. with

```python
for uuid in reader.iter_node_uuids():
   folder = reader.node_repository(uuid)
```

you are looping through `handle.namelist()` for every uuid (but still only performing the same amount of extractions).

The original implementation unpacked the entire zip folder in one loop.
This is more efficient for a full import, but has the massive drawback for general reads, whereby anytime you want to retrieve a single piece of data or a single node repo from the archive, you have to unpack the whole archive.

Could possibly add a "bulk extract" method, but not quite sure how this would work with the abstract interface and also the progress reporter.

Reply:
maybe have a `prepare_for_import` method, which is described as "Prepare the archive for a full import". This would be optional to call and implementation specific (defaulting to doing nothing), but in this case would perform the full unpacking in one loop

Reply:
I might also add a `count_objects` method that e.g. for zip just returns the length of `handle.namelist()`. Might be nice for introspection.

Reply:
> is a bit of a performance regression

Although actually, if you are re-importing an archive, it's a lot more performant, because you only have to unpack new nodes 🤷 (this is probably the less usual case though?)

Reply:
> this, together with the `node_repository` method, is a bit of a performance regression for the current archive format

I suggest you first check how much of a performance regression we're talking about here.
If we're talking about 10% that's one thing - if we're talking about 100% or more that's another.

You could try e.g. the archive from here: https://archive.materialscloud.org/record/2020.107

It might make a difference whether this is done on a SSD or a magnetic disk - I can do a test on a magnetic disk later if you like.

> Although actually, if you are re-importing an archive, it's a lot more performant, because you only have to unpack new nodes 🤷 (this is probably the less usual case though?)

It is a less usual case but it is still relevant (not so much re-importing the same archive, but importing an updated archive that contains a lot of data that was already imported).

Reply:
I've actually fixed this now (commit coming soon..), to change `node_repository` to `iter_node_repos` and it first computes the common prefix of all the required node repositories and only extracts that; so performant for both the single and many node use cases 😄